From 5293ee585f876169e31f3b6af737a42d30302f71 Mon Sep 17 00:00:00 2001
From: Benjamin Berg <bberg@redhat.com>
Date: Wed, 28 Apr 2021 16:50:03 +0200
Subject: [PATCH] gdm: Work around failing fingerprint auth

On Fedora we have the problem that fingerprint auth fails immediately if
the PAM configuration has not been updated and no prints are enrolled.

So, consider a verification failure within one second to be a service
failure instead.
---
 js/gdm/authServicesLegacy.js | 44 +++++++++++++++++++++++++-----------
 1 file changed, 31 insertions(+), 13 deletions(-)

diff --git a/js/gdm/authServicesLegacy.js b/js/gdm/authServicesLegacy.js
index 0813622a01..c85a9572f5 100644
--- a/js/gdm/authServicesLegacy.js
+++ b/js/gdm/authServicesLegacy.js
@@ -58,6 +58,7 @@ export class AuthServicesLegacy extends AuthServices {
         this._addCredentialManager(Vmware.SERVICE_NAME, Vmware.getVmwareCredentialsManager());
 
         this._fingerprintReadyTimeoutId = 0;
+        this._fprintStartTime = -1;
     }
 
     _handleSelectChoice(serviceName, key) {
@@ -117,16 +118,20 @@ export class AuthServicesLegacy extends AuthServices {
     }
 
     _handleOnConversationStarted(serviceName) {
-        if (serviceName === Const.FINGERPRINT_SERVICE_NAME &&
-            this._fingerprintReadyTimeoutId === 0) {
-            this._fingerprintReadyTimeoutId = GLib.timeout_add(
-                GLib.PRIORITY_DEFAULT,
-                FINGERPRINT_READY_TIMEOUT_MS,
-                () => {
-                    this._fingerprintReadyTimeoutId = 0;
-                    this._setFingerprintReady();
-                    return GLib.SOURCE_REMOVE;
-                });
+        if (serviceName === Const.FINGERPRINT_SERVICE_NAME) {
+            this._fprintStartTime = GLib.get_monotonic_time();
+
+
+            if (this._fingerprintReadyTimeoutId === 0) {
+                this._fingerprintReadyTimeoutId = GLib.timeout_add(
+                    GLib.PRIORITY_DEFAULT,
+                    FINGERPRINT_READY_TIMEOUT_MS,
+                    () => {
+                        this._fingerprintReadyTimeoutId = 0;
+                        this._setFingerprintReady();
+                        return GLib.SOURCE_REMOVE;
+                    });
+            }
         }
     }
 
@@ -234,6 +239,7 @@ export class AuthServicesLegacy extends AuthServices {
 
             this._fingerprintFailedId = GLib.timeout_add(GLib.PRIORITY_DEFAULT,
                 FINGERPRINT_ERROR_TIMEOUT_WAIT, () => {
+                    log("Generating _verificationFailed!");
                     this._fingerprintFailedId = 0;
                     if (!this._cancellable.is_cancelled())
                         this._verificationFailed(serviceName, false);
@@ -314,9 +320,21 @@ export class AuthServicesLegacy extends AuthServices {
 
     _handleVerificationFailed(serviceName) {
         if (serviceName === Const.FINGERPRINT_SERVICE_NAME &&
-            this._enabledMechanisms.some(m => m.serviceName === serviceName) &&
-            this._fingerprintFailedId)
-            GLib.source_remove(this._fingerprintFailedId);
+            this._enabledMechanisms.some(m => m.serviceName === serviceName)) {
+            if (this._fingerprintFailedId)
+                GLib.source_remove(this._fingerprintFailedId);
+
+            // On Fedora we have the problem that fingerprint auth fails
+            // immediately if the PAM configuration has not been updated and no
+            // prints are enrolled.
+            // So, consider a verification failure within one second to be a service
+            // failure instead.
+            if (this._fprintStartTime > GLib.get_monotonic_time() - GLib.USEC_PER_SEC) {
+                log("Fingerprint service failed almost immediately, considering it unavailable.");
+                log("Please fix your configuration by running: authselect select --force sssd with-fingerprint with-silent-lastlog");
+                this._onServiceUnavailable(serviceName, null);
+            }
+        }
     }
 
     _handleOnVerificationComplete(serviceName) {
-- 
2.53.0

